couchdb/tests/server/compiler.py

```diff
@@ -99,6 +132,80 @@ def test_fail_for_relative_path_against_root_module(self):
                           './foo/bar/baz'.split('/'), {}, module)
 
+    def test_cache_bytecode_for_future_usage(self):
+        ddoc = {'foo': {'bar': {'baz': 'exports["answer"] = 42'}}}
+        require = compiler.require(ddoc)
+        exports = require('foo/bar/baz')
+        self.assertEqual(exports['answer'], 42)
+        self.assertTrue(isinstance(ddoc['foo']['bar']['baz'], types.CodeType))
+        exports = require('foo/bar/baz')
+        self.assertEqual(exports['answer'], 42)
+
+    def test_cache_egg_exports(self):
+        ddoc = {
+            'lib': {
+                'universe': DUMMY_EGG
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/universe')
+        self.assertEqual(exports['universe'].question.get_answer(), 42)
+        self.assertTrue(isinstance(ddoc['lib']['universe'], dict))
+        exports = require('lib/universe')
+        self.assertEqual(exports['universe'].question.get_answer(), 42)
+
+    def test_reqire_egg_from_module(self):
+        ddoc = {
+            '_id': 'foo',
+            'lib': {
+                'egg': DUMMY_EGG,
+                'utils.py': (
+                    "universe = require('lib/egg')['universe'] \n"
+                    "exports['title'] = 'The Answer' \n"
+                    "exports['body'] = str(universe.question.get_answer())")
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/utils.py')
+        result = ' - '.join([exports['title'], exports['body']])
+        self.assertEqual(result, 'The Answer - 42')
+
+    def test_required_modules_has_global_namespace_access(self):
+        ddoc = {
+            '_id': 'foo',
+            'lib': {
+                'egg': DUMMY_EGG,
+                'utils.py': (
+                    "import math\n"
+                    "def foo():\n"
+                    "  return math.cos(0)\n"
+                    "exports['foo'] = foo")
+            }
+        }
+        require = compiler.require(ddoc, enable_eggs=True)
+        exports = require('lib/utils.py')
+        self.assertEqual(exports['foo'](), 1)
+
+    def test_fail_on_resolving_deadlock(self):
+        ddoc = {
+            'lib': {
+                'stuff': (
+                    "exports['utils'] = require('./utils') \n"
+                    "exports['body'] = 'doc forever!'"),
+                'helper': (
+                    "exports['title'] = 'best ever' \n"
+                    "exports['body'] = require('./stuff')"),
+                'utils': (
+                    "def help():\n"
+                    "  return require('./helper') \n"
+                    "stuff = help()\n"
+                    "exports['title'] = stuff['title'] \n"
+                    "exports['body'] = stuff['body']")
+            }
+        }
+        require = compiler.require(ddoc)
+        self.assertRaises(exceptions.Error, require, 'lib/utils')
```

Comment:
-1. This makes impossible relative requirements from CommonJS modules causing different behaviour with JS QS.